test/jdk/jdk/crac/recompiler/RecompilationDelayTest.java

```diff
@@ -0,0 +1,133 @@
+ * @library /test/lib
+ * @build jdk.test.whitebox.WhiteBox
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller -jar wb.jar jdk.test.whitebox.WhiteBox
+ * @build RecompilationDelayTest
+ * @run driver jdk.test.lib.crac.CracTest -1
+ * @run driver jdk.test.lib.crac.CracTest 0
+ * @run driver jdk.test.lib.crac.CracTest 10000
+ */
+public class RecompilationDelayTest implements CracTest {
+    @CracTestArg(0)
+    long delayMs;
+
+    @SuppressWarnings("unused")
+    private static void testMethod() {}
+
+    private static final String TEST_METHOD_NAME = "testMethod";
+    private static final String MAKE_TEST_METHOD_COMP_BLOCKING_DIRECTIVE =
+        """
+        [
+            {
+                match: "%s.%s",
+                BackgroundCompilation: false
+            }
+        ]
+        """.formatted(RecompilationDelayTest.class.getName().replace('.', '/'), TEST_METHOD_NAME);
+    private static final int TEST_METHOD_COMP_LEVEL = 4;
+
+    @Override
+    public void test() throws Exception {
+        new CracBuilder().engine(CracEngine.SIMULATE)
+            .vmOption("-Xbootclasspath/a:wb.jar").vmOption("-XX:+UnlockDiagnosticVMOptions").vmOption("-XX:+WhiteBoxAPI")
+            .javaOption("jdk.crac.recompilation-delay-ms", Long.toString(delayMs))
+            .startCheckpoint().waitForSuccess();
+    }
+
+    @Override
+    public void exec() throws Exception {
+        final var whiteBox = WhiteBox.getWhiteBox();
+        final var testMethodRef = RecompilationDelayTest.class.getDeclaredMethod(TEST_METHOD_NAME);
+
+        assertEquals(
+            1, whiteBox.addCompilerDirective(MAKE_TEST_METHOD_COMP_BLOCKING_DIRECTIVE),
+            "Unexpected number of directives installed"
+        );
+
+        assertTrue(
+            whiteBox.enqueueMethodForCompilation(testMethodRef, TEST_METHOD_COMP_LEVEL),
+            "Failed to compile immediately"
+        );
+        assertEquals(
+            TEST_METHOD_COMP_LEVEL, whiteBox.getMethodCompilationLevel(testMethodRef),
+            "Unexpected pre-C/R compilation level"
+        );
+
+        final var resource = new Resource() {
+            public long restoreFinishTimeMs = -1;
+
+            @Override
+            public void beforeCheckpoint(Context<? extends Resource> context) throws Exception {}
+
+            @Override
+            public void afterRestore(Context<? extends Resource> context) throws Exception {
+                if (delayMs <= 0) {
+                    assertEquals(
+                        1, whiteBox.deoptimizeMethod(testMethodRef),
+                        "Unexpected number of deoptimizations"
+                    );
+                } else {
+                    restoreFinishTimeMs = Math.floorDiv(System.nanoTime(), 1_000_000);
+                }
+            }
+        };
+        Core.getGlobalContext().register(resource);
+
+        Core.checkpointRestore();
+
+        if (delayMs > 0) {
+            assertEquals(
```

Comment:
10 seconds sounds like enough, but ideally I would have tests as independent to timing as possible. Are there any timestamps in the compilation log so rather than checking the current time we could just read the logs and run assertions on those?

Reply:
I don't think this particular assert can be removed even if we would read logs.

It asserts that the machine had time to deoptimize the method before the delay expired. Because if we do it after the delay expires it is expected that there will be no recompilation. Reading logs won't free us from this assert.

The only thing I can come up with is to implement some mechanism to relaunch the test (or just the restore part if I implement the request above to make recompilation delay restore-settable) with a higher delay when the test fails this way.

Reply:
> So this is actually triggering the deoptimization

Yes, it triggers the deoptimization. It returns the amount of compilations actually deoptimized (can be zero if the method was not compiled; can be multiple for one method if there were OSR compilations with different entry byte codes, although I believe this overload skips OSRs).

> but why wouldn't you trigger it there even for delayMs > 0

Because then the test would succeed even if the delaying code gets removed, i.e. it won't actually be testing that the delay works.

Reply:
It would be easier to test if after-restore recompilations would happen immediately instead of after the delay expires (then I could set a very large delay but after triggering the decompilation just wait a bit until it gets recompiled instead of having to wait until the delay expires). But that turned out to not be easy to implement (threads triggering decompilations sometimes cannot enqueue compilations), so it is something to maybe implement in the future.